### PR TITLE
[airbyte-cdk] - Adds Discover Error Handling

### PR DESCRIPTION
**Pull request by pnilan**
## What
- Adds discover error handler interface and concrete implementations for file-based and default sources.
    - FileBasedDiscoverErrorHandler checks if schema inference or validation error and then ignores stream and logs error, enabling discovery to continue.
    - DefaultDiscoverErrorHandler raises exception and maintains existing behavior for non-file-based connectors
- Ref: https://github.com/airbytehq/airbyte-internal-issues/issues/9187 

## How
- Adds try/catch in `AbstractSource` `discover` method around `stream.as_airbyte_stream()` to catch exceptions, then invokes the source's new discover error handler to determine if the exception should be raised or be logged, enabling the source to continue discovering streams.
- New `AbstractSource` public method `discover_error_handler` initalizes the DiscoverErrorHandler configured for the source.
- Discover errors are now wrapped in `AirbyteTracedExceptions`

## Review guide
1. entrypoint.py
2. abstract_source.py
3. discover_error_handler.py
4. file_based_discover_error_handler.py
5. Remainder

## User Impact
- Users file-based syncs will no longer fail if a user places an incorrect file type in a stream's bucket

## Note
Follow-up issue is to provide a better user experience by informing users that streams were skipped in a sync due to a discovery error. Currently, if the stream is undiscovered, the stream will not be displayed on Cloud. Data will persist in the destination during this scenario and the next time the stream is discovered syncs will resume for the given stream(s).

## Can this PR be safely reverted and rolled back?

- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/entrypoint.py [hunk 1]
```diff
@@ -157,8 +157,17 @@ def discover(self, source_spec: ConnectorSpecification, config: TConfig) -> Iter
...
-        catalog = self.source.discover(self.logger, config)
-
+        try:
+            catalog = self.source.discover(self.logger, config)
+        except AirbyteTracedException as traced_exc:
+            raise traced_exc
+        except Exception as exc:
+            raise AirbyteTracedException(
+                internal_message=f"An error occurred while discovering the source schema. Please check the logged errors for more information: {exc}",
+                failure_type=FailureType.config_error,
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/entrypoint.py:167]: how do you know this is a config error?

File: airbyte-cdk/python/airbyte_cdk/entrypoint.py [hunk 2]
```diff
@@ -169,8 +178,17 @@ def read(self, source_spec: ConnectorSpecification, config: TConfig, catalog: An
 
         # The Airbyte protocol dictates that counts be expressed as float/double to better protect against integer overflows
         stream_message_counter: DefaultDict[HashableStreamDescriptor, float] = defaultdict(float)
-        for message in self.source.read(self.logger, config, catalog, state):
-            yield self.handle_record_counts(message, stream_message_counter)
+        try:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/entrypoint.py:181]: nit: can you separate this change since it's not directly related to to the discover error handling?

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        exceptions_to_log = (InvalidSchemaError, SchemaInferenceError, ConfigValidationError)
+
+        if isinstance(exception, exceptions_to_log):
+            logger.error(
+                f"Error occurred while discovering stream and therefore stream will not be added to the configured catalog: {exception}",
+                exc_info=True,
+            )
+            return None
+        else:
+            return exception
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py:25]: could we raise the exception instead of returning it [just to task the source to raise it](https://github.com/airbytehq/airbyte/pull/44798/files#diff-51c861828e0a614ef6ee21390f3ba17d7138549f0575051fdb24868ee05eabbbR87-R88)?

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    def handle_discover_error(self, logger: logging.Logger, exception: Exception) -> Optional[Exception]:
+        """
+        Default File Based source implementation of the discover error handler. Logs the error or raises the exception.
+        """
+        exceptions_to_log = (InvalidSchemaError, SchemaInferenceError, ConfigValidationError)
+
+        if isinstance(exception, exceptions_to_log):
+            logger.error(
+                f"Error occurred while discovering stream and therefore stream will not be added to the configured catalog: {exception}",
+                exc_info=True,
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py:21]: why the exc_info?

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+from airbyte_cdk.sources.file_based.exceptions import ConfigValidationError, InvalidSchemaError, SchemaInferenceError
+from airbyte_cdk.sources.utils.discover_error_handler import AbstractDiscoverErrorHandler
+
+
+class FileBasedDiscoverErrorHandler(AbstractDiscoverErrorHandler):
+    def handle_discover_error(self, logger: logging.Logger, exception: Exception) -> Optional[Exception]:
+        """
+        Default File Based source implementation of the discover error handler. Logs the error or raises the exception.
+        """
+        exceptions_to_log = (InvalidSchemaError, SchemaInferenceError, ConfigValidationError)
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py:16]: nit: 

We can make the class more generic by passing the types of exceptions to log in the constructor

```
class LogExpectedErrorsHandler(AbstractDiscoverErrorHandler):
    def __init__(exceptions_to_log: List[...]):
       self._exceptions_to_log = exceptions_to_log
   ...
```

Then in the filebased source:
```
    def discover_error_handler(self) -> AbstractDiscoverErrorHandler:
        return FileBasedDiscoverErrorHandler((InvalidSchemaError, SchemaInferenceError, ConfigValidationError))
```

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py [hunk 6]
```diff
@@ -76,7 +76,7 @@ def compute_slices(self) -> Iterable[Optional[Mapping[str, Any]]]:
         slices = [{"files": list(group[1])} for group in itertools.groupby(sorted_files_to_read, lambda f: f.last_modified)]
         return slices
 
-    def read_records_from_slice(self, stream_slice: StreamSlice) -> Iterable[AirbyteMessage]:
+    def read_records_from_slice(self, stream_slice: StreamSlice) -> Iterable[AirbyteMessage]:  # type: ignore # file based stream returns an AirbyteMessage while other implementations return mapping
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py:79]: `read_records_from_slice` only seems to exist for file based connectors. can we align the returned so it always return either AirbyteMessage or Mapping?


```
git grep read_records_from_slice | cat
airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/abstract_file_based_stream.py:        return self.read_records_from_slice(stream_slice)
airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/abstract_file_based_stream.py:    def read_records_from_slice(self, stream_slice: StreamSlice) -> Iterable[Mapping[str, Any]]:
airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/adapters.py:    def read_records_from_slice(self, stream_slice: StreamSlice) -> Iterable[Mapping[str, Any]]:
airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/adapters.py:        yield from self._legacy_stream.read_records_from_slice(stream_slice)
airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py:    def read_records_from_slice(self, stream_slice: StreamSlice) -> Iterable[AirbyteMessage]:
airbyte-cdk/python/unit_tests/sources/file_based/stream/test_default_file_based_stream.py:    def test_when_read_records_from_slice_then_return_records(self) -> None:
airbyte-cdk/python/unit_tests/sources/file_based/stream/test_default_file_based_stream.py:        messages = list(self._stream.read_records_from_slice({"files": [RemoteFile(uri="uri", last_modified=self._NOW)]}))
airbyte-cdk/python/unit_tests/sources/file_based/stream/test_default_file_based_stream.py:    def test_given_exception_when_read_records_from_slice_then_do_process_other_files(
airbyte-cdk/python/unit_tests/sources/file_based/stream/test_default_file_based_stream.py:            self._stream.read_records_from_slice(
airbyte-cdk/python/unit_tests/sources/file_based/stream/test_default_file_based_stream.py:    def test_given_traced_exception_when_read_records_from_slice_then_fail(
airbyte-cdk/python/unit_tests/sources/file_based/stream/test_default_file_based_stream.py:                self._stream.read_records_from_slice(
airbyte-cdk/python/unit_tests/sources/file_based/stream/test_default_file_based_stream.py:    def test_given_exception_after_skipping_records_when_read_records_from_slice_then_send_warning(
airbyte-cdk/python/unit_tests/sources/file_based/stream/test_default_file_based_stream.py:            self._stream.read_records_from_slice(
```

File: airbyte-cdk/python/airbyte_cdk/sources/utils/discover_error_handler.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        """
+        Handles exceptions that occur during the discover process. This method should an exception if the error is unrecoverable. Override this method to implement custom error handling logic.
+        """
+        pass
+
+
+class DefaultDiscoverErrorHandler(AbstractDiscoverErrorHandler):
+    def handle_discover_error(self, logger: logging.Logger, exception: Exception) -> Optional[Exception]:
+        """
+        Default implementation of the discover error handler. Logs the error and raises the exception.
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/utils/discover_error_handler.py:21]: the default implementation doesn't log the error or raise the exception. it's only returned.

I do think that raising the exception is a good default though

File: airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py [hunk 8]
```diff
@@ -246,5 +246,5 @@ def make_file(path: Path, file_contents: Optional[Union[Mapping[str, Any], List[
 
 def get_error_message_from_exc(exc: ExceptionInfo[Any]) -> str:
     if isinstance(exc.value, AirbyteTracedException):
-        return exc.value.message
+        return exc.value.internal_message
```
Comment (girarda) [inline airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py:249]: why?

File: airbyte-cdk/python/airbyte_cdk/entrypoint.py (hunk 1 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/entrypoint.py:167]: Good catch. Updated to `system_error`.

File: airbyte-cdk/python/airbyte_cdk/entrypoint.py (hunk 2 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/entrypoint.py:181]: @girarda 
 TLDR: `read` for file-based sources can invoke `get_json_schema` which I updated from wrapping `SchemaInferenceError` in an `AirbyteTracedException` to just raising it so it can get caught by the new error handling -- so I moved wrapping it to `entrypoint` `read`. Let me know if you don't like this approach.

Longer version:
`entrypoint` `read` invokes the `FileBasedSource` `streams` method which *can* (if concurrency_level is not `None) result in initializing the `FileBasedStreamFacade` which in turns initializes `DefaultStream` with a json_schema argument set to `stream.get_json_schema()` which no longer wraps the `SchemaInferenceError` in an ATE.

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py (hunk 3 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py:25]: I think so. We can implement another type of error handler if we want different handling

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py (hunk 4 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py:21]: Fair point. I think that ends up being redundant with including the exception in the log message. I can remove.

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py (hunk 6 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py:79]: Updated.

File: airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py (hunk 8 above)
Comment (pnilan) [reply airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py]: Reverted

File: airbyte-cdk/python/airbyte_cdk/entrypoint.py (hunk 2 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/entrypoint.py]: After discussion, we've decided to revert this and more specifically, the way `InvalidSchemaError` is handled within `default_file_based_stream` `get_json_schema` method. It now gets wrapped in an `AirbyteTracedException` like before, and the `file_based_discover_error_handler` now checks for the underlying exception if an `AirbyeTracedException` is being handled.

File: airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py [hunk 9]
```diff
@@ -67,11 +68,30 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
...
+                streams.append(stream.as_airbyte_stream())
+            except Exception as exception:
+                errors.append(exception)
+                self.discover_error_handler().handle_discover_error(logger=logger, exception=exception)
+
+        if not streams:
+            raise AirbyteTracedException(
+                internal_message=f"No streams were discovered in the source. Please check the logged errors for more information: {errors}",
+                message="No streams were discovered in the source. Please check the source configuration.",
+                failure_type=FailureType.config_error,
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py:92]: open question: should we always raise a config error or should it be 
- file-based: config_error
- else: system_error

the reasoning being that I wouldn't expect other connectors to not have any streams discovered

Implementation wise, this might just mean adding a method that returns a boolean on the discover error handler

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        ] or []
+
+    def handle_discover_error(self, logger: logging.Logger, exception: Exception) -> None:
+        if type(exception) in self._exceptions_to_log:
+            logger.error(
+                f"Error occurred while discovering stream and therefore stream will not be added to the configured catalog: {exception}"
+            )
+        elif (
+            isinstance(exception, AirbyteTracedException)
+            and exception.message is not None
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/error_handlers/file_based_discover_error_handler.py:30]: can we use the AirbyteTracedException's exception field instead of the error message? It seems more reliable, and it would simplify the interface since `underlying_exceptions_to_log` would be a list of error types instead of a union[error type, str]

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/abstract_file_based_stream.py [hunk 11]
```diff
@@ -101,7 +101,7 @@ def read_records(
         return self.read_records_from_slice(stream_slice)
 
     @abstractmethod
-    def read_records_from_slice(self, stream_slice: StreamSlice) -> Iterable[Mapping[str, Any]]:
+    def read_records_from_slice(self, stream_slice: StreamSlice) -> Union[Iterable[AirbyteMessage], Iterable[Mapping[str, Any]]]:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/abstract_file_based_stream.py:104]: are you changing the type because it was already wrong or because we're not returning AirbyteMessages in addition to mappings?

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/adapters.py [hunk 12]
```diff
@@ -229,7 +229,7 @@ def read(self) -> Iterable[Record]:
                     data_to_return = dict(record_data)
                     self._stream.transformer.transform(data_to_return, self._stream.get_json_schema())
                     yield Record(data_to_return, self.stream_name())
-                elif isinstance(record_data, AirbyteMessage) and record_data.type == Type.RECORD:
+                elif isinstance(record_data, AirbyteMessage) and record_data.type == Type.RECORD and record_data.record is not None:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/adapters.py:232]: in what scenario would a record have no data? I don't think adding it to the message repository would be the right thing to do when that happens

File: airbyte-cdk/python/airbyte_cdk/sources/utils/discover_error_handler.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+from abc import ABC, abstractmethod
+
+
+class AbstractDiscoverErrorHandler(ABC):
+    """Abstract class for handling exceptions that occur during the discover process."""
+
+    @abstractmethod
+    def handle_discover_error(self, logger: logging.Logger, exception: Exception) -> None:
+        """
+        Handles exceptions that occur during the discover process. This method should an exception if the error is unrecoverable. Override this method to implement custom error handling logic.
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/utils/discover_error_handler.py:14]:
```suggestion
        Handles exceptions that occur during the discover process. This method should an raise exception if the error is unrecoverable. Override this method to implement custom error handling logic.
```
